### DISCUSSION TITLE
How to convert a .mo file of OpenModelica to a .mex file or a .dll file to be used in MATLAB?

### DISCUSSION THREAD

**Discussion by OMOCMAT2024** [OpenModelica/OpenModelica#13278] [Q&A]
Hello all,

I am new to OpenModelica and I have no experience in compiling .mex or .dll file that MATLAB can use. I have a simple .mo file like this:

model SimpleModel
    parameter Real a = 1.0; 
    Real x(start = 1.0); 
    input Real u; 
equation
    der(x) = -a * x + u;
end SimpleModel;

I want to convert it into either a .mex or .dll file so that I can use MATLAB to simulate this model. I used the following commands in OM Shell to generate some .c and .h files:

loadFile("SimpleModel.mo")
instantiateModel(SimpleModel)
translateModel(SimpleModel)

However, I don't know how to proceed next. I also tried to convert the .mo file into an FMU, and I saw there is a .dll file inside the binary folder in FMU, but I don't know how I can use that with MATLAB.

Can anyone give me some guidance of converting such a .mo file into a .mex or .dll file that can be used in MATLAB for simulation? Thanks in advance.

P.S. I have both OpenModelica v1.24.2 and MATLAB 2021a in Windows 10.

**Comment by adrpo** [comment]
Export as FMU then import FMU in Matlab. 
See https://fmi-standard.org/

Right click on the model and say Export in the left hand browser in OMEdit (OpenModelica Connection Editor)

**Reply by OMOCMAT2024** [reply]
Thank you for your reply. I exported the .mo file as FMU but I don't know how I can use it with MATLAB. I know Simulink can import FMU as a block but I only want to use with MATLAB (script). There is a commerical FMI toolbox from Modelon https://modelon.com/fmi-toolbox/ but I do not intent to use it. 

Due to .mo file can be compiled into C code and .dll file by OpenModelica, and MATLAB can compile C code into .mex file or use .dll file, I am looking for a such a method.

If you know how to important FMU directly into MATLAB (scirpt) and carry out a simulation, please let me know the steps too. Thanks in advance.

**Reply by adrpo** [reply]
I do not know if what you want is possible.

**Reply by adrpo** [reply]
It seems you can import the FMU via scripting:
https://se.mathworks.com/matlabcentral/answers/711398-fmu-import-script-for-simulink

**Reply by OMOCMAT2024** [reply]
Thank you for your reply. The FMU via scripting works with Simulink still. I want to avoid Simulink and only works on MATALB. I think comipling the generated C code via .mo file into the .mex file in MATLAB is possible, since the mex function in MATLAB is for C code compiling. I just do not know the detailed steps to make it work. I can generate the C code of a .mo file by:

loadFile("SimpleModel.mo")
instantiateModel(SimpleModel)
translateModel(SimpleModel)

But I do not know how to proceed next.

**Reply by adrpo** [reply]
The answer is not that simple, the C code needs to be linked with a solver that integrates time. I know nothing about mex, it might be possible but you need to know the internal details of the Matlab and the OpenModelica runtime.

**Reply by OMOCMAT2024** [reply]
Is that possible to generate the C code on the model part only without any solver? Indeed the generated C code by default has to include of a lot of libraries and so on, which is difficult to understand and use MATLAB to comiple them.

**Comment by casella** [comment]
@OMOCMAT2024 please check the [FMI toolbox](https://www.mathworks.com/products/connections/product_detail/modelon-fmi-toolbox.html) by Modelon. The documentation claims that you can access the FMU from command line and scripts (i.e., MATLAB), without the need of using Simulink.

**Reply by OMOCMAT2024** [reply]
Thank you for the advice. I want to use any open-source or free approach to use FMU if possible.

**Reply by casella** [reply]
MATLAB is not open-source nor free 😃 

That said, Matlab allows to import FMUs natively into Simulink. You can then use the MATLAB functions to interact with a Simulink model from a script.

You could also try to wrap the FMU into an S-function by writing suitable C code, but you'd need to do it yourself and I'm not sure to which extent that would be possible.

**Reply by adrpo** [reply]
Just a small question: why do you need Matlab?

**Reply by OMOCMAT2024** [reply]
Thank you for the advice. Just for computing efficiency, if Matlab can import FMU directly, why do we have to go through Simulink? @adrpo mainly for simulation and optimization purpose, would be nice to have Matlab working with FMU directly. Is there a Python tool that work with FMU? If so, how I can work in Python with FMU for simulation and optimization? Thanks!

**Reply by adrpo** [reply]
I guess you might be able to use FMPy via the Python Matlab interface? 

https://github.com/CATIA-Systems/FMPy

**Reply by casella** [reply]
> Thank you for the advice. Just for computing efficiency, if Matlab can import FMU directly

I'm not sure it can. The FMI toolbox provides functions to do that, but I guess there's nothing like that natively in Matlab.

> mainly for simulation and optimization purpose, would be nice to have Matlab working with FMU directly. I
You need to ask MathWorks, we can't do anything about it 

>is there a Python tool that work with FMU? 

Yes, there are two of them: FMPy and PyFMI. We are using FMPy for our FMU testing, works quite well. 

> If so, I can work in Python with FMU for simulation and optimization? 

For simulation, you can use FMPy .For optimization, check Pyomo, though I don't think it can use FMIs, it needs explicit symbolic representations of the model.

**Reply by OMOCMAT2024** [reply]
@adrpo thanks, I will have a look. @casella I am looking at Pyomo, looks like it is an interface to formulate genearl optimization problems. Does FMU provide symbolic Jacobians of the model inside the FMU? I am looking for a possible workflow like this: OpenModelica --> FMU (with symbolic Jacobian) --> Python (or Matlab) based optimization tool for dynamic optimization (ideally to use the symbolic Jacobian within the FMU).

**Reply by casella** [reply]
> @casella I am looking at Pyomo, looks like it is an interface to formulate genearl optimization problems. Does FMU provide symbolic Jacobians of the model inside the FMU?

The Jacobians may be computed by code that used automatic differentation, but you only get numerical values from an FMU, no symbolic information, except the sparsity structure.

One possible workflow is OpenModelica --> Base Modelica flat output --> Pyomo.DAE input. @matteodepascali has been experimenting successfully with that approach, though it has some limitations, the most important being:
- the model must be index-1
- the model must be entirely equation-based, no Modelica functions that cannot be inlined and no external C functions

We plan to publish about that and to make the Python code that implements the workflow open source in the next few months.

**Reply by OMOCMAT2024** [reply]
Hello @casella , is automatic differentiation embedded in FMU to compute Jacobians? Or external automatic differentation or any other external derivative approximation methods can access the FMU to compute Jacobians? I heard tools like Dymola can generate Jaocibans automatically for their use in optimization like in MATLAB (maybe through .mex or .dll files?), are those Jacobians generated by Dymola symbolic and evaluated numerically in MATLAB? Can similar mechanism being used in the FMU by OpenModelica?

The main reason I want to use OpenModelica is that its capabilities in things like indexing reduction and optimizing DAE systems and generating (symoblic) Jacobians, which I read from literatures. If both the function bodies of the DAE system as well as its Jacobian can be evaluated with a given FMU, then they can be provided to a solver such as IPOPT for optimization. If only funciton bodies of the DAE but not the Jaociban can be evaluated of an FMU, how can an external automatic differentaion method evaluate the Jacobians of the model of an FMU? (I don't want to use finite difference due to its inaccuracy). 

The transcription for a continuous DAE to an NLP I have done in MATLAB. I know Pyomo.DAE was developed for this purpose as well. At the moment I do not know how I can get FMU to work with either MATLAB or Pyomo to get its Jacobians.

I read limitation of index-1 from papers regarding how (Open)Modelica optimizes DAEs and generates (symoblic) Jacobians (use in OpenModelica simulation). In optimal control, usually there are variable bound contraints and inequality path constraints, they make a high index DAEs anyway I guess? Would you recommend to only formulate a model using OpenModelica with ODEs or equality path constraints can be formulated with ODEs together? For the second limiation, an entirely equation-based model is OK for me, since that was how I formulate an optimal control problem before.

Thank you for your advice.

**Reply by casella** [reply]
> Hello @casella , is automatic differentiation embedded in FMU to compute Jacobians? 

Yes, if you turn on --generateSymbolicJacobians, which became --generateDynamicJacobians = symbolic in 1.25.0. 

> Or external automatic differentation or any other external derivative approximation methods can access the FMU to compute Jacobians?

No. FMU functions just return numbers, not formulae.

>  I heard tools like Dymola can generate Jaocibans automatically for their use in optimization like in MATLAB (maybe through .mex or .dll files?), are those Jacobians generated by Dymola symbolic and evaluated numerically in MATLAB?

I guess so.

> Can similar mechanism being used in the FMU by OpenModelica?

See above.

> The main reason I want to use OpenModelica is that its capabilities in things like indexing reduction and optimizing DAE systems and generating (symoblic) Jacobians, which I read from literatures. If both the function bodies of the DAE system as well as its Jacobian can be evaluated with a given FMU

Note that the FMU contains the system reduced to state-space form, i.e., explicit ODEs plus output equations

> then they can be provided to a solver such as IPOPT for optimization.

In principle, yes. Check Ruediger Franke's publications.

>  If only funciton bodies of the DAE but not the Jaociban can be evaluated of an FMU, how can an external automatic differentaion method evaluate the Jacobians of the model of an FMU? (I don't want to use finite difference due to its inaccuracy).

It can't, as far as I understand.

> The transcription for a continuous DAE to an NLP I have done in MATLAB.

Good. Then, you can look into [Base Modelica](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/basemodelica.html#array-preserving-basemodelica-output) flat output..

> I know Pyomo.DAE was developed for this purpose as well. At the moment I do not know how I can get FMU to work with either MATLAB or Pyomo to get its Jacobians.

There are API function in the FMI standard to get the (numerical values of the) Jacobians, check it out. Tools like FMPy can also read the xml file for structural information, so they know which variables show up in which equations, allowing for efficient sparse Jacobian computation and use.

> I read limitation of index-1 from papers regarding how (Open)Modelica optimizes DAEs and generates (symoblic) Jacobians (use in OpenModelica simulation). In optimal control, usually there are variable bound contraints and inequality path constraints, they make a high index DAEs anyway I guess?

No. The problem is that equality constraints given by high-index DAEs lead to singular problems when transcribed. Try what happens with, e.g., 

```modelica
der(x1) = 1 - y;
der(x2) = y;
x1 = 3*x2;
```

> Would you recommend to only formulate a model using OpenModelica with ODEs or equality path constraints can be formulated with ODEs together? For the second limiation, an entirely equation-based model is OK for me, since that was how I formulate an optimal control problem before.

I see no need to write ODEs in OpenModelica. Go for DAEs. The problem is only if DAEs are index-1. If you export the model as an FMU, it is first subject to symbolic index reduction to turn it into state-space form (that's what an FMU-ME is, in essence). If you export the Base Modelica flat model, index reduction is not performed, so you need to take care of it yourself.

We are considering how to get index reduction into the Base Modelica export process, but it's not easy because currently the flat output is provided by the front-end and the index reduction is performed by the back-end, which use different data structures, so it's not easy to back-port index-reduced models into the front-end for flat output.

Hope this helps.

**Reply by OMOCMAT2024** [reply]
Hi @casella , thank you so much for such detailed replies. I need some time to look for the information you provided, and I do not think I can understand them immediately. I will reply you later in more details. Again thank you very much.
